### DISCUSSION TITLE
I have small task on Ton blockchain.

### DISCUSSION THREAD

**Discussion by xxargonaut** [community/community#135556] [Programming Help] [answered]
### Body

Update the contract from below code so that it has the following functionality:
- store in c4 the address of the sender of the last message and the integer total - a 32-bit unsigned number;
- when receiving an internal incoming message, the contract must take the address from in_msg, a 32-bit unsigned integer from the body of the message, add it to total, and store the address and amount in the contract data;
- The smart contract must have a get_sum method that allows you to return the total value;

```
#include "imports/stdlib.fc";

() recv_internal(int msg_value, cell in_msg, slice in_msg_body) impure {
	slice cs = in_msg.begin_parse();
	int flags = cs~load_uint(4);
	slice sender_address = cs~load_msg_addr();

	set_data(begin_cell().store_slice(sender_address).end_cell());
}

slice get_the_latest_sender() method_id {
   slice ds = get_data().begin_parse();
   return ds~load_msg_addr();
}
```

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by everforyou** [comment, marked as answer]
Hi @xxargonaut 

### Updated Contract Code

you need to modify the existing contract to include the following functionalities:

1. Store the address of the sender and the total in a 32-bit unsigned integer.
2. Update the total when receiving an internal message.
3. Implement a `get_sum` method to return the total value.

like this

```c
#include "imports/stdlib.fc";

cell c4;

() recv_internal(int msg_value, cell in_msg, slice in_msg_body) impure {
    slice cs = in_msg.begin_parse();
    int flags = cs.load_uint(4);
    slice sender_address = cs.load_msg_addr();    
    uint total = get_data().begin_parse().load_uint(32);    
    uint amount = cs.load_uint(32);    
    total += amount;
    set_data(begin_cell().store_slice(sender_address).store_uint(total, 32).end_cell());
}

slice get_sum() method_id {
    slice ds = get_data().begin_parse();
    return ds.load_uint(32);
}
```

- **recv_internal**: This functions handles incoming messages. It retrieved the sender's address and the amount from the message, updates the total, and stores both in the contract's state.
- **get_sum**: This method extracts the total from the stored data and returns it,.

**Comment by 0xpuck** [comment]
Hi
Here is the updated code
```
#include "imports/stdlib.fc";

() recv_internal(int msg_value, cell in_msg, slice in_msg_body) impure {
    slice cs = in_msg.begin_parse();
    int flags = cs~load_uint(4);
    slice sender_address = cs~load_msg_addr();
    
    int amount = in_msg_body~load_uint(32);


    slice ds = get_data().begin_parse();
    int current_total = ds~load_uint(32);

    int new_total = current_total + amount;

    set_data(
        begin_cell()
            .store_uint(new_total, 32)
            .store_slice(sender_address)
        .end_cell()
    );
}

int get_sum() method_id {
    slice ds = get_data().begin_parse();
    return ds~load_uint(32);
}
```

**Comment by xxargonaut** [comment]
Thank you for your considerations
